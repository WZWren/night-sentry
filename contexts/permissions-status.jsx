export const LocalPermStatus = Object.freeze({
    INIT: 'init',
    GRANTED: 'granted',
    UNDETERMINED: 'undetermined',
    DENIED: 'denied',
});