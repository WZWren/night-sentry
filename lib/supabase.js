import 'react-native-url-polyfill/auto'
import { createClient } from "@supabase/supabase-js";
import AsyncStorage from '@react-native-async-storage/async-storage'

const projectUrl = process.env.PROJECT_URL
const projectKey = process.env.PROJECT_KEY

export const supabase = createClient(projectUrl, projectKey, {
    auth: {
        storage: AsyncStorage,
    }
});