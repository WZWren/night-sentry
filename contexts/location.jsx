import { useState, useEffect, createContext, useContext } from 'react';
import * as Location from 'expo-location';

import { LocalPermStatus } from './permissions-status';

const LocationContext = createContext({});

export function useLocation() {
    return useContext(LocationContext);
}

export function LocationProvider({children}) {
    const [permissionStatus, setPermissionStatus] = useState(null);
    const [location, setLocation] = useState(null);

    useEffect(() => {
        if (permissionStatus == LocalPermStatus.INIT) {
            (async () => {
                let { status } = await Location.requestForegroundPermissionsAsync();
                setPermissionStatus(status);
                if (status !== LocalPermStatus.GRANTED) {
                  console.log('Permission to access location was denied/not accepted yet.');
                  return;
                }
                let location = await Location.getCurrentPositionAsync({});
                setLocation(location);
                console.log(location.coords.latitude + ", " + location.coords.longitude);
            })();
        }
    }, [permissionStatus]);

    return (
        <LocationContext.Provider value={{ location, permissionStatus, setPermissionStatus }}>
            {children}
        </LocationContext.Provider>
    );
}