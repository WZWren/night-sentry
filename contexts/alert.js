/**
 * alert.js manages all the PubSub stuff for alerts and services.
 * Functions in this js file will be called back in auth during the login step.
 */

import { supabase } from "../lib/supabase";

export async function fetchPublishers(user, setDistress) {
    const { data, error } = await supabase.from("close_contacts")
        .select("publisher, name:publisher(first_name, last_name)")
        .eq("subscriber", user.id)
        .is("confirmed", true);

    if (error) {
        console.log(error.message);
        return;
    }

    // linear runtime is OK for this, as a user's close contacts should
    // on average be within the 10s range.
    for (let i = 0; i < data.length; i++) {
        // subscribe the app client to the channel.
        supabase.channel(data[i].publisher)
            .on('postgres_changes',
                { event: 'INSERT', schema: 'public', table: 'alerts', filter: `user_id=eq.${data[i].publisher}` },
                (payload) => {
                    setDistress(`Distress signal received from ${data[i].name.first_name} ${data[i].name.last_name}`);
                    console.log(payload);
                }
            ).subscribe();
        console.log("Subscribed to " + data[i].publisher)
    }
}

export async function unsubClient() {
    const { error } = await supabase.removeAllChannels();

    if (error) {
        console.log(error.message);
        return;
    }
    console.log("unsubscribed to all");
}