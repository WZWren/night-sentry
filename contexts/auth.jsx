import { useRouter, useSegments } from "expo-router";

import { createContext, useContext, useEffect, useState } from "react";
import { supabase } from "../lib/supabase";

const AuthContext = createContext({});

function useProtectedRoute(loggedIn) {
    const segments = useSegments();
    const router = useRouter();

    useEffect(() => {
        console.log("Entering protected route...");
        // this checks if the route is in the (auth) group of routes - in case
        // the user is somehow not on this route if the user is not logged in
        const inAuthGroup = segments[0] === "(auth)";
        if (!loggedIn && !inAuthGroup) {
            console.log("Routing to Login...");
            router.replace("/login");
        } else if (loggedIn && inAuthGroup) {
            console.log("Routing to Alert...");
            router.replace("/alert");
        }
    }, [loggedIn, segments, router]);
}

/**
 * Parent component that provides the user value as part of its context.
 * 
 * @param {*} children - The Child Components of the AuthProvider
 * @returns The AuthContext.Provider component, encapsulating the user value.
 */
export function AuthProvider({ children }) {
    const [loggedIn, setLoggedIn] = useState(null);
    console.log("AuthProvider loading...");

    useProtectedRoute(loggedIn);

    // this sets up the useEffect hooks to redirect users to the correct page.
    useEffect(() => {
        const { data } = supabase.auth.onAuthStateChange((event, session) => {
            console.log("User changed! " + event);
            if (event === "SIGNED_IN") {
                setLoggedIn(session.user);
            } else if (event === "SIGNED_OUT") {
                setLoggedIn(null);
            }
        })
        return () => data.subscription.unsubscribe();
    }, []);

    return (
        <AuthContext.Provider value={{ loggedIn }}>{ children }</AuthContext.Provider>
    );
}