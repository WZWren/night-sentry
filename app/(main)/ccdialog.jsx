import { useState } from "react";
import { Dialog, Portal, TextInput, Button } from "react-native-paper";
import { useAuth } from "../../contexts/auth";
import { supabase } from "../../lib/supabase";

async function fetchEmail(user, email, setVisible) {
    const { data, error: emailError } = await supabase.from("user_info").select("id").eq("email", email);
    if (emailError) {
        console.log(emailError.message);
        return;
    } else if (data.length <= 0) {
        console.log("No such user found: " + data);
        return;
    }

    const { error: insertError } = await supabase.from("close_contacts").insert({
        subscriber: user.id,
        publisher: data[0].id,
        confirmed: false
    });
    if (insertError) {
        console.log(insertError.message);
    }
    setVisible(false);
}

export function CCDialog({ visible, setVisible }) {
    const { loggedIn } = useAuth();
    const [ email, setEmail ] = useState("");

    return (
        <Portal>
            <Dialog visible={ visible } onDismiss={ () => setVisible(false) }>
                <Dialog.Title>Add a close contact...</Dialog.Title>
                <Dialog.Content>
                    <TextInput
                        placeholder="Enter Close Contact Email"
                        autoCapitalize="none"
                        autoCorrect={false}
                        textContentType="emailAddress"
                        value={email}
                        onChangeText={setEmail} />
                </Dialog.Content>
                <Dialog.Actions>
                    <Button onPress={() => fetchEmail(loggedIn, email, setVisible)}>Submit</Button>
                </Dialog.Actions>
            </Dialog>
        </Portal>
    )
}