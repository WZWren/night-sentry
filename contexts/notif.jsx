// TODO: Move Notifications to its separate context.

import { createContext, useContext, useEffect, useState, useRef } from "react";
import { Platform } from "react-native";
import * as Device from 'expo-device';
import * as Notifications from 'expo-notifications';

import { supabase } from "../lib/supabase";
import { useAuth } from "./auth";

const NotificationsContext = createContext({});

export function useNotif() {
    return useContext(NotificationsContext);
}

// we handle the notification signup in the layout slot.
// notifications taken directly from the expo-notifications example.
Notifications.setNotificationHandler({
    handleNotification: async () => ({
        shouldShowAlert: true,
        shouldPlaySound: true,
        shouldSetBadge: false,
    }),
});

async function registerForPushNotificationsAsync() {
    let token;

    if (Platform.OS === 'android') {
        await Notifications.setNotificationChannelAsync('default', {
            name: 'default',
            importance: Notifications.AndroidImportance.MAX,
            vibrationPattern: [0, 250, 250, 250],
            lightColor: '#FF231F7C',
        });
    }

    if (Device.isDevice) {
        const { status: existingStatus } = await Notifications.getPermissionsAsync();
        let finalStatus = existingStatus;
        if (existingStatus !== 'granted') {
            const { status } = await Notifications.requestPermissionsAsync();
            finalStatus = status;
        }
        if (finalStatus !== 'granted') {
            console.log('Failed to get push token for push notification!');
            return;
        }
        token = (await Notifications.getExpoPushTokenAsync()).data;
        console.log(token);
    } else {
        console.log('Must use physical device for Push Notifications');
    }

    return token;
}

export function NotificationsProvider({ children }) {
    const { loggedIn } = useAuth();
    // expoPushToken is to be pushed to the supabase user_info table.
    const [ expoPushToken, setExpoPushToken ] = useState("");
    // notification is the physical notification from the app
    const [ notification, setNotification ] = useState(null);
    const notificationListener = useRef();
    const responseListener = useRef();

    useEffect(() => {
        registerForPushNotificationsAsync().then(token => setExpoPushToken(token));
    
        notificationListener.current = Notifications.addNotificationReceivedListener(notification => {
            setNotification(notification);
            console.log("Notification received! " + notification.request.content.title);
        });
    
        responseListener.current = Notifications.addNotificationResponseReceivedListener(response => {
            // TODO: this is where router linking will take place.
            console.log(response);
        });
    
        return () => {
            Notifications.removeNotificationSubscription(notificationListener.current);
            Notifications.removeNotificationSubscription(responseListener.current);
        };
    }, []);

    useEffect(() => {
        if (loggedIn && expoPushToken) {
            console.log("Updating user token: " + expoPushToken);
            (async () => {
                const { error } = await supabase.from("user_info")
                    .update({ expo_token: expoPushToken })
                    .eq("id", loggedIn.id);
                if (error) {
                    console.log(error.message);
                }
            })();
        }
    }, [loggedIn, expoPushToken]);

    return (
        <NotificationsContext.Provider value={{ notification, setNotification, expoPushToken }}>
            {children}
        </NotificationsContext.Provider>
    );
}