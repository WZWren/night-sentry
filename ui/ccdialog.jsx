import { useState } from "react";
import { Dialog, Portal, TextInput, Button, Text, ActivityIndicator } from "react-native-paper";
import { useAuth } from "../contexts/auth";
import { supabase } from "../lib/supabase";
import { View } from "react-native";

/**
 * Asynchronously checks the email input to see if it is a valid user - if it is, sends a request to
 * the user.
 * 
 * @param {*} user The currently logged-in user.
 * @param {*} email The email of the close contact to add.
 * @param {*} setVisible Setter for the hook to set visibility of the dialog.
 * @param {*} setLoading Setter for the UI loading bar element.
 * @param {*} setError Setter for the error message to show the user.
 */
async function handleNewCC(user, email, setVisible, setLoading, setError) {
    if (email == "") {
        setError("Fields should not be empty.");
        return;
    }
    setLoading(true);
    setError("");
    const { data, error: emailError } = await supabase.from("user_info").select("id").eq("email", email);
    if (emailError) {
        setError(emailError.message);
        setLoading(false);
        return;
    } else if (data.length <= 0) {
        setError("No such user found!");
        setLoading(false);
        return;
    }

    const { error: insertError } = await supabase.from("close_contacts").insert({
        subscriber: user.id,
        publisher: data[0].id,
        confirmed: false
    });
    if (!insertError) {
        setLoading(false);
        setVisible(false);
        return;
    }
    switch(insertError.code) {
        case "23505":
            setError("Request for close contact already exists.");
            break;
        case "23514":
            setError("Cannot be a close contact of yourself.");
            break;
        default:
            setError(`${insertError.message}`);
    }
    setLoading(false);
}

/**
 * Dialog box for adding close contacts to the user.
 * 
 * @param {*} {visible} Hook for visibility. Required for dialog box implementation.
 * @param {*} {setVisible} Allows the dialog box to close itself internally.
 */
export function CCDialog({ visible, setVisible }) {
    const { loggedIn } = useAuth();
    const [ email, setEmail ] = useState("");
    const [ loading, setLoading ] = useState(false);
    const [ error, setError ] = useState("");

    return (
        <Portal>
            <Dialog visible={ visible } onDismiss={ () => setVisible(false) }>
                <Dialog.Title>Add a close contact...</Dialog.Title>
                <Dialog.Content>
                    <TextInput
                        placeholder="Enter Close Contact Email"
                        autoCapitalize="none"
                        autoCorrect={false}
                        textContentType="emailAddress"
                        value={email}
                        onChangeText={setEmail}
                        style={{ margin: 4 }}/>
                    <View style={{ minHeight: 30, alignItems: "center"}}>
                        {!loading && <Text variant="labelLarge" testID="error_msg">{error}</Text>}
                        {loading && <ActivityIndicator/>}
                    </View>
                </Dialog.Content>
                <Dialog.Actions>
                    <Button onPress={() => handleNewCC(loggedIn, email, setVisible, setLoading, setError)}>Submit</Button>
                </Dialog.Actions>
            </Dialog>
        </Portal>
    )
}