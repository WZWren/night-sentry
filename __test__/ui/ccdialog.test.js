import { screen, render, fireEvent, act } from "@testing-library/react-native";
import { rest } from "msw";
import { setupServer } from "msw/node";

import 'react-native-url-polyfill/auto';
import { createClient } from "@supabase/supabase-js";

import { CCDialog } from "../../ui/ccdialog";
import * as Auth from "../../contexts/auth";
import * as DB from "../../lib/supabase";
import { PaperProvider } from "react-native-paper";

const URL = process.env.PROJECT_URL;
const KEY = process.env.PROJECT_KEY;

// this mirrors the internal implementation of resolveFetch for Supabase.
var _fetch = async (...args) => await (await import('cross-fetch')).fetch(...args);
const fetcher = (...args) => _fetch(...args);

const replacement = createClient(URL, KEY, {
    auth: {
        persistSession: false,
    },
    global: {
        fetch: fetcher,
    }
});

const server = setupServer(
    rest.post(`${URL}/rest/v1/close_contacts`, async (req, res, ctx) => {
        const reqJson = await req.json();
        if (reqJson.publisher == reqJson.subscriber) {
            // if the publisher = subscriber, the backend will respond with an error code.
            return res(
                ctx.status(400),
                ctx.json({
                    "code": "23514",
                    "details": "Failing row contains -snip-.",
                    "hint": null,
                    "message": "new row for relation \"close_contacts\" violates check constraint \"close_contacts_check\""
                })
            );
        } else {
            // otherwise it will successfully insert.
            return res(
                ctx.status(201)
            );
        }
    })
);

// NOTE: As the only test that should calls setVisible is the one at the back, we make no reset
// attempt on this mock. Should more tests be necessary, a afterEach handler to reset the mock
// as cleanup should be created.
const setVisible = jest.fn();

beforeAll(() => {
    jest.useFakeTimers();
    server.listen();
    jest.replaceProperty(DB, "supabase", replacement);
    jest.spyOn(Auth, "useAuth").mockImplementation(() => ({
        loggedIn: {
            // this is a randomly generated uuid not attached to any users.
            id: "aaa46c03-6d2a-4090-ace5-835808756c86"
        }
    }));
});

afterAll(() => {
    server.close();
    jest.useRealTimers();
    jest.restoreAllMocks();
});

it("should render properly", async () => {
    const renderPage = render(
        <PaperProvider>
            <CCDialog visible={true} setVisible={setVisible}/>
        </PaperProvider>
    );
    const submitBtn = await renderPage.findByText("Submit");
    const errMsg = await renderPage.findByTestId("error_msg");
    expect(submitBtn).toBeTruthy();
    expect(errMsg.children[0]).toBeFalsy();

    const tree = renderPage.toJSON();
    expect(tree).toMatchSnapshot();
    renderPage.unmount();
});

it("should not call Supabase API if field is empty", async () => {
    const spy = jest.spyOn(DB.supabase, "from");
    const renderPage = render(
        <PaperProvider>
            <CCDialog visible={true} setVisible={setVisible}/>
        </PaperProvider>
    );
    const submitBtn = await renderPage.findByText("Submit");
    await act(() => {
        fireEvent.press(submitBtn);
    });

    const errMsg = await renderPage.findByTestId("error_msg");
    expect(spy).not.toBeCalled();
    expect(errMsg.children[0]).toBe("Fields should not be empty.");
    renderPage.unmount();
    spy.mockReset();
});

it("should resolve an error if a non-user email is supplied", async () => {
    server.use(
        rest.get(`${URL}/rest/v1/user_info`, (req, res, ctx) => {
            // on the backend, the get request returns nothing if email is not in DB.
            return res.once(
                ctx.status(200),
                ctx.json([])
            );
        })
    );

    const renderPage = render(
        <PaperProvider>
            <CCDialog visible={true} setVisible={setVisible}/>
        </PaperProvider>
    );

    const inputField = await renderPage.findByPlaceholderText("Enter Close Contact Email");
    const submitBtn = await renderPage.findByText("Submit");
    fireEvent.changeText(inputField, "placeholder");
    fireEvent.press(submitBtn);

    const errMsg = await renderPage.findByTestId("error_msg");
    expect(errMsg.children[0]).toBe("No such user found!");
    renderPage.unmount();
});

it("should resolve an error on insertion attempt failure", async () => {
    // we use the non-reflexivity relation of close-contact for this, as it is the easiest to
    // spoof.
    server.use(
        rest.get(`${URL}/rest/v1/user_info`, (req, res, ctx) => {
            // on the backend, the get request returns nothing if email is not in DB.
            return res.once(
                ctx.status(200),
                ctx.json([{
                    "id": "aaa46c03-6d2a-4090-ace5-835808756c86"
                }])
            );
        })
    );

    const renderPage = render(
        <PaperProvider>
            <CCDialog visible={true} setVisible={setVisible}/>
        </PaperProvider>
    );

    const inputField = await renderPage.findByPlaceholderText("Enter Close Contact Email");
    const submitBtn = await renderPage.findByText("Submit");
    fireEvent.changeText(inputField, "placeholder");
    fireEvent.press(submitBtn);

    const errMsg = await renderPage.findByTestId("error_msg");
    // we just need some error message to be made, since ccdialog handles the error with a
    // default error message if it is uncaught.
    expect(errMsg.children[0]).toBeTruthy();
    // we also expect that the dialog should not attempt to close on failure.
    expect(setVisible).not.toBeCalled();
    renderPage.unmount();
});

it("should attempt to close the dialog on success", async () => {
    server.use(
        rest.get(`${URL}/rest/v1/user_info`, (req, res, ctx) => {
            // on the backend, the get request returns nothing if email is not in DB.
            return res.once(
                ctx.status(200),
                ctx.json([{
                    "id": "cc17eedf-4060-415e-b2a9-4366626f2349"
                }])
            );
        })
    );

    const renderPage = render(
        <PaperProvider>
            <CCDialog visible={true} setVisible={setVisible}/>
        </PaperProvider>
    );

    const inputField = await renderPage.findByPlaceholderText("Enter Close Contact Email");
    const submitBtn = await renderPage.findByText("Submit");
    fireEvent.changeText(inputField, "placeholder");
    fireEvent.press(submitBtn);

    const errMsg = await renderPage.findByTestId("error_msg");
    // we expect the errMsg to be blank.
    expect(errMsg.children[0]).toBeFalsy();
    // we also expect that the dialog to close.
    expect(setVisible).toBeCalled();
    renderPage.unmount();
});

// handler for get requests - supabase seems to encrypt its search params, so we define the interceptor as runtime
// this is the backup code for making sense of the logic.
/*
rest.get(`${URL}/rest/v1/user_info`, (req, res, ctx) => {
    console.log("intercepted!");
    console.log(req);
    if (req.params.email === "nonsense") {
        // on the backend, the get request returns nothing if email is not in DB.
        return res(
            ctx.status(200),
            ctx.json([])
        );
    } else if (req.params.email === "self") {
        return res(
            ctx.status(200),
            ctx.json([{
                "id": "aaa46c03-6d2a-4090-ace5-835808756c86"
            }])
        );
    } else { // this is other email that is not self that is valid.
        return res(
            ctx.status(200),
            ctx.json([{
                "id": "cc17eedf-4060-415e-b2a9-4366626f2349"
            }])
        );
    }
}),
*/