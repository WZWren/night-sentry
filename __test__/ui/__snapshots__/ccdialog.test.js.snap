// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render properly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  />
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      accessibilityLiveRegion="polite"
      accessibilityViewIsModal={true}
      collapsable={false}
      onAccessibilityEscape={[Function]}
      pointerEvents="auto"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="modal"
    >
      <View
        accessibilityLabel="Close modal"
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        importantForAccessibility="no"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "rgba(50, 47, 55, 0.4)",
            "flex": 1,
            "opacity": 1,
          }
        }
        testID="modal-backdrop"
      />
      <View
        pointerEvents="box-none"
        style={
          [
            {
              "bottom": 0,
              "justifyContent": "center",
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            {
              "marginBottom": 0,
              "marginTop": 0,
            },
            undefined,
          ]
        }
        testID="modal-wrapper"
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "rgb(238, 232, 244)",
              "borderRadius": 28,
              "marginHorizontal": 26,
              "marginVertical": 0,
              "opacity": 1,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 1,
                "width": 0,
              },
              "shadowOpacity": 0.15,
              "shadowRadius": 3,
            }
          }
          testID="modal-surface-outer-layer"
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "rgb(238, 232, 244)",
                "borderRadius": 28,
                "elevation": 24,
                "flex": undefined,
                "justifyContent": "flex-start",
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 1,
                  "width": 0,
                },
                "shadowOpacity": 0.3,
                "shadowRadius": 1,
              }
            }
            testID="modal-surface"
          >
            <Text
              accessibilityRole="header"
              style={
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "fontFamily": "System",
                      "fontSize": 24,
                      "fontWeight": "400",
                      "letterSpacing": 0,
                      "lineHeight": 32,
                    },
                    [
                      {
                        "marginBottom": 18,
                        "marginHorizontal": 24,
                        "marginTop": 22,
                      },
                      {
                        "marginBottom": 16,
                        "marginTop": 16,
                      },
                      {
                        "color": "rgba(28, 27, 31, 1)",
                        "fontFamily": "System",
                        "fontSize": 24,
                        "fontWeight": "400",
                        "letterSpacing": 0,
                        "lineHeight": 32,
                      },
                      [
                        {
                          "marginTop": 24,
                        },
                        undefined,
                      ],
                    ],
                  ],
                ]
              }
            >
              Add a close contact...
            </Text>
            <View
              style={
                [
                  {
                    "paddingBottom": 24,
                    "paddingHorizontal": 24,
                  },
                  undefined,
                ]
              }
            >
              <View
                style={
                  [
                    {
                      "backgroundColor": "rgba(231, 224, 236, 1)",
                      "borderTopLeftRadius": 4,
                      "borderTopRightRadius": 4,
                    },
                    {
                      "margin": 4,
                    },
                  ]
                }
              >
                <View
                  collapsable={false}
                  style={
                    {
                      "backgroundColor": "rgba(28, 27, 31, 1)",
                      "bottom": 0,
                      "height": 1,
                      "left": 0,
                      "position": "absolute",
                      "right": 0,
                      "transform": [
                        {
                          "scaleY": 0.5,
                        },
                      ],
                      "zIndex": 1,
                    }
                  }
                  testID="text-input-underline"
                />
                <View
                  style={
                    [
                      {
                        "paddingBottom": 0,
                        "paddingTop": 0,
                      },
                      {
                        "minHeight": 56,
                      },
                    ]
                  }
                >
                  <TextInput
                    autoCapitalize="none"
                    autoCorrect={false}
                    cursorColor="rgba(103, 80, 164, 1)"
                    editable={true}
                    maxFontSizeMultiplier={1.5}
                    multiline={false}
                    onBlur={[Function]}
                    onChangeText={[Function]}
                    onFocus={[Function]}
                    placeholder="Enter Close Contact Email"
                    placeholderTextColor="rgba(73, 69, 79, 1)"
                    selectionColor="rgba(103, 80, 164, 1)"
                    style={
                      [
                        {
                          "margin": 0,
                        },
                        {
                          "paddingLeft": 16,
                          "paddingRight": 16,
                        },
                        {
                          "height": 56,
                        },
                        {
                          "paddingBottom": 0,
                          "paddingTop": 0,
                        },
                        {
                          "color": "rgba(73, 69, 79, 1)",
                          "fontFamily": "System",
                          "fontSize": 16,
                          "fontWeight": undefined,
                          "letterSpacing": 0.15,
                          "lineHeight": undefined,
                          "textAlign": "left",
                          "textAlignVertical": "center",
                        },
                        false,
                        [
                          {},
                        ],
                        undefined,
                      ]
                    }
                    testID="text-input-flat"
                    textContentType="emailAddress"
                    underlineColorAndroid="transparent"
                    value=""
                  />
                </View>
              </View>
              <View
                style={
                  {
                    "alignItems": "center",
                    "minHeight": 30,
                  }
                }
              >
                <Text
                  style={
                    [
                      {
                        "textAlign": "left",
                      },
                      {
                        "color": "rgba(28, 27, 31, 1)",
                        "writingDirection": "ltr",
                      },
                      [
                        {
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        undefined,
                      ],
                    ]
                  }
                  testID="error_msg"
                />
              </View>
            </View>
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "flexGrow": 1,
                    "justifyContent": "flex-end",
                    "paddingBottom": 24,
                    "paddingHorizontal": 24,
                  },
                  undefined,
                ]
              }
            >
              <View
                collapsable={false}
                style={
                  {
                    "backgroundColor": "transparent",
                    "borderRadius": 20,
                    "marginRight": 0,
                    "shadowColor": "#000",
                    "shadowOffset": {
                      "height": 0,
                      "width": 0,
                    },
                    "shadowOpacity": 0,
                    "shadowRadius": 0,
                  }
                }
                testID="button-container-outer-layer"
              >
                <View
                  collapsable={false}
                  style={
                    {
                      "backgroundColor": "transparent",
                      "borderColor": "transparent",
                      "borderRadius": 20,
                      "borderStyle": "solid",
                      "borderWidth": 0,
                      "flex": undefined,
                      "minWidth": "auto",
                      "shadowColor": "#000",
                      "shadowOffset": {
                        "height": 0,
                        "width": 0,
                      },
                      "shadowOpacity": 0,
                      "shadowRadius": 0,
                    }
                  }
                  testID="button-container"
                >
                  <View
                    accessibilityRole="button"
                    accessibilityState={
                      {
                        "busy": undefined,
                        "checked": undefined,
                        "disabled": false,
                        "expanded": undefined,
                        "selected": undefined,
                      }
                    }
                    accessibilityValue={
                      {
                        "max": undefined,
                        "min": undefined,
                        "now": undefined,
                        "text": undefined,
                      }
                    }
                    accessible={true}
                    collapsable={false}
                    focusable={true}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      [
                        {
                          "overflow": "hidden",
                        },
                        {
                          "borderRadius": 20,
                        },
                      ]
                    }
                    testID="button"
                  >
                    <View
                      style={
                        [
                          {
                            "alignItems": "center",
                            "flexDirection": "row",
                            "justifyContent": "center",
                          },
                          undefined,
                        ]
                      }
                    >
                      <Text
                        numberOfLines={1}
                        selectable={false}
                        style={
                          [
                            {
                              "textAlign": "left",
                            },
                            {
                              "color": "rgba(28, 27, 31, 1)",
                              "writingDirection": "ltr",
                            },
                            [
                              {
                                "fontFamily": "System",
                                "fontSize": 14,
                                "fontWeight": "500",
                                "letterSpacing": 0.1,
                                "lineHeight": 20,
                              },
                              [
                                {
                                  "marginHorizontal": 16,
                                  "marginVertical": 9,
                                  "textAlign": "center",
                                },
                                false,
                                {
                                  "marginHorizontal": 12,
                                },
                                {
                                  "marginHorizontal": 8,
                                },
                                false,
                                {
                                  "color": "rgba(103, 80, 164, 1)",
                                  "fontFamily": "System",
                                  "fontSize": 14,
                                  "fontWeight": "500",
                                  "letterSpacing": 0.1,
                                  "lineHeight": 20,
                                },
                                undefined,
                              ],
                            ],
                          ]
                        }
                        testID="button-text"
                      >
                        Submit
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
